File: bindings/wasm/poi_wasm/examples/src/03_reuse_verifier_for_multiple_proofs.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * # Reuse a Verifier for Multiple Proofs
 *
 * A verifier retains resolved committees in memory, so applications should
 * reuse it across proofs from the same network.
 */

import { fromBase58 } from "@iota/iota-sdk/utils";
import {
    createNotarization,
    elapsedMilliseconds,
    loadGenesisCommitteeResolution,
    preparePoiExample,
} from "./util.js";

/** Demonstrates how one verifier reuses committee resolution across proofs. */
export async function reuseVerifierForMultipleProofs(): Promise<void> {
    console.log("=== Proof of Inclusion: Reuse a Verifier for Multiple Proofs ===\n");

    console.log("Stage 1 - Configure the proof source and establish committee trust");
    const context = await preparePoiExample();
    const trust = await loadGenesisCommitteeResolution(context);
    console.log(`  trust anchor: ${trust.description}`);

    console.log("\nStage 2 - Create two Notarization objects using Locked Notarization");
    const firstTargets = await createNotarization(context, "First Notarization object");
    const secondTargets = await createNotarization(context, "Second Notarization object");

    console.log("\nStage 3 - Construct one transaction proof for each Notarization object");
    const firstProof = await context.poiClient.makeProof({
        transaction: fromBase58(firstTargets.transactionDigest),
    });
    const secondProof = await context.poiClient.makeProof({
        transaction: fromBase58(secondTargets.transactionDigest),
    });

    console.log(`  first proof epoch:  ${firstProof.checkpointEpoch}`);
    console.log(`  second proof epoch: ${secondProof.checkpointEpoch}\n`);

    // Keep this verifier alive so its committee cache can serve both proofs.
    const verifier = context.poiClient.verifier(trust.resolution);

    console.log("Stage 4 - Verify both proofs with one verifier");
    console.log("  verifying the first proof; this resolves its checkpoint committee...");
    const firstStart = process.hrtime.bigint();
    await verifier.verify(firstProof);
    const firstDuration = elapsedMilliseconds(firstStart);

    console.log("  verifying the second proof with the same verifier...");
    const secondStart = process.hrtime.bigint();
    await verifier.verify(secondProof);
    const secondDuration = elapsedMilliseconds(secondStart);

    console.log("\n  both transaction proofs verified successfully.");
    console.log(`  first verification:  ${firstDuration.toFixed(2)} ms`);
    console.log(`  second verification: ${secondDuration.toFixed(2)} ms`);
    console.log(`Both proofs used committee resolution through ${trust.description}.`);
}

File: examples/poi/03_reuse_verifier_for_multiple_proofs.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! # Reuse a Verifier for Multiple Proofs
//!
//! Genesis-anchored verification authenticates every committee transition from
//! epoch 0 to the proof's checkpoint epoch. A verifier retains each authenticated
//! committee in its cache, so applications should reuse it across proofs from the
//! same network instead of creating a fresh verifier for every proof.
//!
//! This example creates two transactions on the active network and constructs a
//! proof for each. The second verification reuses committee history already
//! authenticated while verifying the first proof.

use std::time::Instant;

use anyhow::{Context, Result};
use poi_examples::prepare_poi_example;
use poi_rs::CommitteeResolution;

/// Demonstrates how to:
/// 1. Establish committee trust from the active network's genesis blob.
/// 2. Create and construct independent proofs for two transactions.
/// 3. Create one genesis-anchored verifier.
/// 4. Verify the first proof and populate the authenticated committee cache.
/// 5. Reuse the same verifier for the second proof.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Proof of Inclusion: Reuse a Verifier for Multiple Proofs ===\n");

    let context = prepare_poi_example().await?;
    let genesis = context.load_genesis().await?;
    let resolution = CommitteeResolution::from_genesis(genesis)
        .context("failed to load the committee from the trusted genesis blob")?;
    let first_transaction = context
        .create_notarization("PoI verifier-reuse example, transaction 1")
        .await?
        .transaction_digest;
    let second_transaction = context
        .create_notarization("PoI verifier-reuse example, transaction 2")
        .await?
        .transaction_digest;
    let client = &context.poi_client;

    // Each builder creates an independent portable proof.
    let first_proof = client
        .proof()
        .transaction(first_transaction)
        .build()
        .await
        .context("failed to construct the first transaction proof")?;
    let second_proof = client
        .proof()
        .transaction(second_transaction)
        .build()
        .await
        .context("failed to construct the second transaction proof")?;

    println!("Network:            {}", context.network_alias);
    println!("First transaction:  {first_transaction}");
    println!(
        "First checkpoint:   {}",
        first_proof.checkpoint_summary().sequence_number
    );
    println!("First epoch:        {}", first_proof.checkpoint_summary().epoch());
    println!("Second transaction: {second_transaction}");
    println!(
        "Second checkpoint:  {}",
        second_proof.checkpoint_summary().sequence_number
    );
    println!("Second epoch:       {}\n", second_proof.checkpoint_summary().epoch());

    // Keep this verifier alive. Its default in-memory cache stores only
    // committees authenticated through the genesis-anchored epoch walk.
    let verifier = client.verifier(resolution);

    println!("Verifying the first proof; this performs the epoch walk...");
    let first_started = Instant::now();
    verifier
        .verify(&first_proof)
        .await
        .context("first transaction proof verification failed")?;
    let first_elapsed = first_started.elapsed();

    println!("Verifying the second proof with the same verifier...");
    let second_started = Instant::now();
    verifier
        .verify(&second_proof)
        .await
        .context("second transaction proof verification failed")?;
    let second_elapsed = second_started.elapsed();

    println!("\nBoth transaction proofs verified successfully.");
    println!("First verification:  {first_elapsed:?}");
    println!("Second verification: {second_elapsed:?}");
    println!("The second verification reused committee history authenticated during the first verification.");

    Ok(())
}
